### ISSUE TITLE
Inference Performance

### ISSUE THREAD

**Issue by tomschelsen** [DigitalPhonetics/IMS-Toucan#64]
FastPitch, FastSpeech2 and Avocodo all claim inference times faster than 100x real time, sometimes way more (FastPitch), on GPU.

When doing inference on an Nvidia A100, I barely get above 1x (2.6 seconds to generate **3 seconds of audio**), measured with python's `time.perf_counter()` on the `FastSpeech2Interface.forward()` function (so model loading or wav file writing are not taken into account).

More precisely, those are the results (which stay roughly the same across several runs) : 
```
Text2phone done in 0.0018 seconds.
Phone2mel done in 0.6225 seconds.
Mel2wav done in 1.8307 seconds.
Enhancement done in 0.1348 seconds.
```

Do other people get similar results or not with different GPUs and what could be the explanation ? Are papers inflating their inference real-case end-to-end performance, is it the implementation (did people compare with other FastSpeech2 / HifiGAN implementations), or is it a bug way lower level like cuda version related ?

I am on the v2.3 release. 

Any help greatly appreciated, thanks :)

**Comment by Flux9665** [comment]
Your numbers look closer to what I get on CPU. It takes me around 4 seconds to generate 5 seconds worth of audio on CPU. On my RTX 2070 in my home PC, generation is almost instant. I can do 20 seconds of audio in significantly less than a second of compute. Are you sure that the models are actually running on the A100?

Other than that, there are a bunch of factors to this: 
- Our vocoder performs super-resolution as well as spectrogram inversion. Generating a 48kHz wave from a 16kHz spectrogram requires a lot more computations than what they were doing in the paper. That probably explains why this is the biggest time-sink in your observations. The new vocoder with a lower resolution should take about half the time.
- They often inflate their numbers by reporting the real time factor you get when running a whole batch at inference time and then summing up the length of all the sequences in the batch. This can be cheated, because if you just use a larger GPU, you can put more sequences in a batch, while taking about the same amount of time. We do one after another during inference because that makes everything a lot simpler and I never found speed to be a problem.
- There are some overheads because of things that happen in inference mode etc. You would probably get a better real-time-factor on a longer sequence than on a shorter sequence.
- Our configurations are close to the papers, but not a perfect match. Most of the code is based on ESPNet and simply wrapped in easier-to-use interfaces. So some deviation is natural.

**Comment by tomschelsen** [comment]
Thanks for the insight. It's strange and... reassuring at the same time that you get better performance on GPU ("several seconds of audio in significantly less than a second of compute" would be fine), which means the architecture is not the bottleneck and that I should be able to get to the same point (I was expecting some deviation from the papers but, as you pointed, not in the CPU-instead-of-GPU performance scale).

Some progress : yesterday's experiments were using `run_text_to_file_reader.py`, which calls `InferenceInterfaces.FastSpeech2Interface.InferenceFastSpeech2`. I switched it to `InferenceInterfaces.ControllableInterface.ControllableInterface`, which for reasons I don't really understand allows to explicitely set a gpu_id and does some initialization that `InferenceFastSpeech2` doesn't : https://github.com/DigitalPhonetics/IMS-Toucan/blob/ee3b7988b15f871414829d30f24b5cc2a9d097b3/InferenceInterfaces/ControllableInterface.py#L15

And that indeed resulted in some differences : Text2phone compute time was consistently divided by 3 and Phone2mel compute time by 10... But the vocoder and enhancement parts stayed the same : 
```
Text2phone done in 0.0005 seconds.
Phone2mel done in 0.0686 seconds.
Mel2wav done in 2.0619 seconds.
Enhancement done in 0.1507 seconds.
```

So... I don't know how to go from there... Some things are definitely loaded on the gpu (using `watch nvidia-smi` I can see the VRAM go from 0 to 10GiB), but I am not familiar with more precise tracing of Nvidia GPU loading/compute. I'll check what tools are available on the infra.

In the mean time would you mind sharing the Toucan/Torch/CUDA/NvidiaDriver versions with which you get the results you described on GPU ? So that I can try to be on the same ground as you (if that's possible, regarding hardware differences).

Thanks a lot.

PS: I can't sadly take advantage of the batching / longer sequence trick as my use case is conversationnal with 10-20 word sentences mainly. More on that use-case if I manage to get good enough performance ;) but it's multilingual, multispeaker, and needs controllability, so a perfect match for Toucan :)

**Comment by Flux9665** [comment]
The InferenceInterface for FastSpeech does take the device as an argument, where you can pass in a torch.device, which can be a specific GPU. We just leave out the part where we make all the other GPUs invisible to avoid accidentially running something from the dependency on another GPU. I might add that part as well, but there should be no difference in speed.

https://github.com/DigitalPhonetics/IMS-Toucan/blob/1c581e06b1cabf07c89568169cf4317aa50b9896/InferenceInterfaces/FastSpeech2Interface.py#L25

I'm not sure about the enhancement, because that is an external model. But it should only be used if you are using extremely low quality data for training and it seems to be relatively quick anyway, so I'm not too concerned.

It is very strange that the mel2wav part is so slow for you though. I don't see a reason this might be in the code. All of the models are definitely loaded to the same device, otherwise you would get an error about the tensor being passed through not being on the same device. There must be some overhead somewhere, but I don't really see where.

Torch version: 1.13.0
CUDA version: 11.7
NvidiaDriver: GameReady 527.37

But I don't think the issue lies in any of those versions. 

> PS: I can't sadly take advantage of the batching / longer sequence trick as my use case is conversationnal with 10-20 word sentences mainly. More on that use-case if I manage to get good enough performance ;) but it's multilingual, multispeaker, and needs controllability, so a perfect match for Toucan :)

Sounds cool, then the upcoming versions of Toucan will be pretty well suited for your use as well!

**Comment by tomschelsen** [comment]
So... I think I found an explanation / solution. I was researching Pytorch optimization in general, and I discovered here https://pytorch.org/tutorials/intermediate/nvfuser_intro_tutorial.html#setup-and-performance-metrics that "warming-up" pytorch was actually a thing.

And so, instead of doing what I did previsously (each run = load the model -> start the timer -> infer ->  stop the timer), I setup a test where I load the model once, then do several runs of inference, measuring the time on each run (so this is not batching I guess, as the input given for the inference is still the same short one every time), which consistently give those interesting results : 
```
Inference done in 2.65740 seconds.
Duration : 4.64 seconds.
RTF : 1.7x
-----

Inference done in 7.84477 seconds.
Duration : 4.848 seconds.
RTF : 0.6x
-----

Inference done in 0.04679 seconds.
Duration : 4.848 seconds.
RTF : 103.6x
-----

Inference done in 0.06031 seconds.
Duration : 4.432 seconds.
RTF : 73.5x
-----

Inference done in 0.05607 seconds.
Duration : 4.832 seconds.
RTF : 86.2x
-----

Inference done in 0.05303 seconds.
Duration : 4.432 seconds.
RTF : 83.6x
-----

Inference done in 0.05192 seconds.
Duration : 4.848 seconds.
RTF : 93.4x
-----

Inference done in 0.05972 seconds.
Duration : 4.528 seconds.
RTF : 75.8x
-----

Inference done in 0.05499 seconds.
Duration : 4.736 seconds.
RTF : 86.1x
-----

Inference done in 0.05099 seconds.
Duration : 4.464 seconds.
RTF : 87.5x
-----

Inference done in 0.05048 seconds.
Duration : 4.624 seconds.
RTF : 91.6x
-----

Inference done in 0.05103 seconds.
Duration : 4.816 seconds.
RTF : 94.4x
-----

Inference done in 0.05106 seconds.
Duration : 4.72 seconds.
RTF : 92.4x
-----

Inference done in 0.05066 seconds.
Duration : 4.784 seconds.
RTF : 94.4x
-----

Inference done in 0.05075 seconds.
Duration : 4.656 seconds.
RTF : 91.7x
-----

Inference done in 0.05062 seconds.
Duration : 4.608 seconds.
RTF : 91.0x
-----

Inference done in 0.04536 seconds.
Duration : 4.528 seconds.
RTF : 99.8x
-----

Inference done in 0.05044 seconds.
Duration : 4.672 seconds.
RTF : 92.6x
-----

Inference done in 0.05038 seconds.
Duration : 4.336 seconds.
RTF : 86.1x
-----

Inference done in 0.05087 seconds.
Duration : 4.48 seconds.
RTF : 88.1x
-----
```

So every time, the first inference gives the bad performance I was observing previously, the second one is even way worse, then all the follwing ones perform adequately (~90x real-time). And I can deal with that (I'll do a dozen warm-up runs after loading the model, and only accept inference from there).

So I am closing the issue, thank you very much for your support in helping to understand what to expect and what were the invariances.

**Comment by Flux9665** [comment]
Interesting, thanks for sharing that explanation! I heard about lazy loading where things would only be loaded the first time they are used and then stay loaded, instead of being loaded immediately when something is initialized. Maybe this is the case because I convert the generator of the vocoder into a compiled version using the trace function of torchscript. That should in theory make it faster, but maybe this is where this overhead on the first use comes from.

It's just strange that I don't have this super slow first two samples, when I run it, it is fast immediately. Hopefully this will all be better when torch2.0 releases and we can just compile everything :)
